// express-mate/dist/ApiFail.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var http_status_1 = __importDefault(require("http-status"));
var Responder_1 = require("./Responder");
var settings_1 = require("./settings");
var ApiFail = /** @class */ (function () {
    function ApiFail(res, data, meta) {
        if (data === void 0) { data = {}; }
        if (meta === void 0) { meta = {}; }
        this.data = data;
        this.meta = meta;
        this.res = res;
    }
    ApiFail.respond = function (res, data, opt) {
        if (opt === void 0) { opt = {}; }
        var _a = opt.responseFormat, responseFormat = _a === void 0 ? settings_1.Settings.responseFormat : _a, meta = opt.meta;
        var instance = new ApiFail(res, data, meta);
        return Responder_1.triggerResponder(instance, responseFormat);
    };
    Object.defineProperty(ApiFail.prototype, "status", {
        get: function () {
            return ApiFail.status;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ApiFail.prototype, "code", {
        get: function () {
            return ApiFail.code;
        },
        enumerable: true,
        configurable: true
    });
    ApiFail.prototype.raw = function () {
        this.res.status(this.code).json(this.data);
    };
    ApiFail.prototype.jsend = function () {
        this.res.status(this.code).json(__assign(__assign({}, this.meta), { status: this.status, data: this.data }));
    };
    ApiFail.status = 'fail';
    ApiFail.code = http_status_1.default.BAD_REQUEST;
    return ApiFail;
}());
exports.ApiFail = ApiFail;
