// express-mate/dist/ApiUnauthorized.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var http_status_1 = __importDefault(require("http-status"));
var Responder_1 = require("./Responder");
var settings_1 = require("./settings");
var ApiUnauthorized = /** @class */ (function () {
    function ApiUnauthorized(res, message, meta) {
        if (message === void 0) { message = 'Unauthorized'; }
        if (meta === void 0) { meta = {}; }
        this.message = message;
        this.meta = meta;
        this.res = res;
    }
    ApiUnauthorized.respond = function (res, message, opt) {
        if (opt === void 0) { opt = {}; }
        var _a = opt.responseFormat, responseFormat = _a === void 0 ? settings_1.Settings.responseFormat : _a, meta = opt.meta;
        var instance = new ApiUnauthorized(res, message, meta);
        return Responder_1.triggerResponder(instance, responseFormat);
    };
    Object.defineProperty(ApiUnauthorized.prototype, "status", {
        get: function () {
            return ApiUnauthorized.status;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ApiUnauthorized.prototype, "code", {
        get: function () {
            return ApiUnauthorized.code;
        },
        enumerable: true,
        configurable: true
    });
    ApiUnauthorized.prototype.raw = function () {
        this.res.status(this.code).send(this.message);
    };
    ApiUnauthorized.prototype.jsend = function () {
        this.res.status(this.code).json(__assign(__assign({}, this.meta), { status: this.status, message: this.message }));
    };
    ApiUnauthorized.status = 'error';
    ApiUnauthorized.code = http_status_1.default.UNAUTHORIZED;
    return ApiUnauthorized;
}());
exports.ApiUnauthorized = ApiUnauthorized;
